pkg/reconciler/revision/reconcile_resources.go

```diff
@@ -117,6 +117,10 @@ func (c *Reconciler) reconcileDeployment(ctx context.Context, rev *v1.Revision)
 		}
 	}
 
+	if *deployment.Spec.Replicas > 0 && *deployment.Spec.Replicas == deployment.Status.ReadyReplicas {
```

Comment:
@SaschaSchwarze0 hi, should we relax the condition since?

We set the container as unhealthy when a permanent failure is found:

```
	// If a container keeps crashing (no active pods in the deployment although we want some)
	if *deployment.Spec.Replicas > 0 && deployment.Status.AvailableReplicas == 0 {
```
Resetting this should happen when we have at least one pod up no?
I am thinking of a scenario where not all pods become ready (eg. some stay in pending state) but then we keep the revision in false status. WDYTH?

Reply:
Hi @skonto, sorry for the late response. I was out for some time.

Can you please help me and clarify what you're after. The code change I am making is to set container healthy to true. You now come up with a discussion and a related condition about when container healthy should be set to false?

Reply:
Hi @SaschaSchwarze0 I am saying that a revision is serving traffic even when pods are not all ready right? So resetting this to true only when `*deployment.Spec.Replicas == deployment.Status.ReadyReplicas` seems a bit too strict no?

Reply:
I see.

Correct, a revision may be serving traffic even if it does have unhealthy containers in some of the replicas.

Whether that means that it is okay to set ContainerHealthy to True if already one of the replicas is running fine, I do not know.

My proposed condition basically was when I think that ContainerHealthy should definitely be set to true (because all replicas are fully ready).

How strict or relaxed we can be, I do not know. At some point I tried to find a spec on the exact meaning of the conditions of the different resources. But I had not found something.

What your condition can btw lead to is the following:

* You have a revision with two replicas, one is not healthy, the other one is.
* The code in https://github.com/knative/serving/blob/v0.42.2/pkg/reconciler/revision/reconcile_resources.go#L86 checks the first pod only. If that one is not healthy, it will set ContainerHealthy=False.
* I think, the code place here runs later. That one will notice that there are two replicas and one is ready, and set ContainerHealthy to True.

Would not happen with my proposed code because if one pod is not ready, there are not all replicas of the deployment ready. Anyway, just checking the first pod to determine whether ContainerHealthy is set to False is another questionable piece of code because it leads to different results when one has different pod statuses depending on the order in which the pods are returned.

Reply:
I agree. Btw I am also trying to figure out the semantics here. cc @dprotaso who may know more

Reply:
The semantics here are terrible, because distributed systems.  Sorry! 

What users want: 

* When my application is able to handle requests, `Ready` should be true
* If `Ready` is false, there's a clear condition as to why
* If a container crashes on startup, users would like to be about to see the exit message

The real world laughs at this: 

* Containers may pass an initial health check and then have a subsequent failure in application code (for example, if a backend dependency becomes unavailable)
* Kubernetes mostly assumes that nodes are interchangeable, but what if pods on one node have different behavior than pods on another node (e.g. network partition)
* There are a bunch of pre-container-executing failures that can happen, like failure to schedule, pull images, mount volumes, etc.  I'm not sure that's relevant here, but worth considering in the big picture 

The end result is that "does a container become ready" is not really deterministic (and we don't have a good diagnostic on whether failures are permanent or transient), so all we have here are heuristics. 

My gut would be relax the check to closer to what Stavros is suggesting: container ready can become true if there is at least one ready container.  While it's trivial to build counter-example containers for this rule, I think it's probably close enough for most users. 

Sorry if all that is unsatisfying...

Reply:
Thanks for the good discussion.

I think from the comments I conclude the following:

1. There is currently no definition on the ContainerHealthy condition of a Revision.
2. We acknowledge that there may be temporary glitches that we do not necessary want ContainerHealthy to turn False.

To address this, based on what I personally think and what I read as opinions from others, would we agree on the following?

1. If there is no Pod, we do not change the current status.
2. If any Pod is ready, we set it to True.
3. If no Pod is ready and if there is at least one non-terminating Pod and if all non-terminating Pods have a container with LastTerminationState.Terminated != nil, then we set it to False.
4. If neither (2) nor (3) apply, we also do not change the current status.

The code for (2) would be simple because it can be evaluated simply by looking at the deployment status:

```go
if deployment.Status.AvailableReplicas > 0 {
  set ContainerHealthy=true
}
```

The code for (3) is more complex as it requires a look at the Pods. I guess we would need to make https://github.com/knative/serving/blob/v0.42.2/pkg/reconciler/revision/reconcile_resources.go#L84-L117 more complex (= not just look at the first pod, though it can break out immediatly if a pod is ready). (2) can probably also be solved here.

This should be safe for scale up from 0 to 1. In that case, we will have one Pod, none are ready, but none have containers with restarts. We would not change the current status until this one Pod crashes.

ScaleDown from 1 to 0 also works. The one Pod may have a restart that can be long ago (OOM or similar), it will not be Ready while it terminates, but because it is terminating, that is ignored.

Wdyt?

Reply:
For reference, more discussion in https://cloud-native.slack.com/archives/C04LMU0AX60/p1733333864902479

Reply:
> I think @skonto is suggesting:
> 
> Is that right?

Yes.  We tag a revision with containerHealthy=false because all pods are down for a specific condition. After that the only thing that can happen is this condition to be reverted to unblock the revision. If the condition is reverted for one pod it implies that it has been reverted for all pods independently of when those pods will become ready again. We just need a signal for the condition not holding any more. Serving should handle traffic as usual depending on the number of pods that are ready at any given time.

There can be the case that some pods are not scheduled yet and we do cover that. If there are no pods (some failure has not created the pods) we do cover that I think since we propagate the deployment status (replicaset failures etc).